### ISSUE TITLE
False Results?

### ISSUE THREAD

**Issue by SrikanthSoma** [speed47/spectre-meltdown-checker#143]
I ran following script on Amazon Linux 2017.09 which is using following kernel version: Linux ip-172-31-9-230 **4.9.76**-3.78.amzn1.x86_64 #1 SMP Fri Jan 12 19:51:35 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux. Script says variant 1(CVE-2017-5715) is vulnerable but amazon documentation says those kernels are fixed 
sources: https://aws.amazon.com/speculative-execution-os-updates/ 
               https://aws.amazon.com/security/security-bulletins/AWS-2018-013/

Ubuntu 14.04 and 16.04
**************************
As per the tables on https://wiki.ubuntu.com/SecurityTeam/KnowledgeBase/SpectreAndMeltdown  -  kernel 3.13 for amd64 systems is protected from variant 1, variant 2 and Meltdown (the i386 version is only fully protected from variant 1 though). 

root@ip-172-31-5-120:~/spectre-meltdown-checker# uname -a
Linux ip-172-31-5-120 3.13.0-141-generic #190-Ubuntu SMP Fri Jan 19 12:52:38 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

The script reported similar errors: the variant 1 check was "unknown", the variant 2 was "vulnerable" and variant 3 was "not vulnerable".

The same issue was diagnosed in 16.04 - which ran the kernel version 4.4.0-1049-aws. This is protected as per the above Ubuntu documentation: "Cloud images created using 4.4.0-108.131 and its derivatives (for example, linux-aws 4.4.0-1047.56) have the mitigations for Meltdown." 4.4.0-1049-aws is higher than 4.4.0-1047.56 and therefore contains the patches, but the script diagnosed this system as vulnerable.

To summarize, I've patched 3 different instances with the latest available package updates offered by the official OS repository - none of them were found to be protected somehow. I would enquire with the script publisher as to what is their methodology of fixing the vulnerabilities detected, since there are no newer patches available for the issue the script detects; alternatively - are these false-positive results?

**Comment by speed47** [comment]
Thanks for your report.
I've been working on a branch to enhance detection for Canonical(Ubuntu) and RedHat versions of the patches, that are different from the upstream Linux kernel patches.
Could you try it again with the redhat_canonical branch and report the results here ? ( https://raw.githubusercontent.com/speed47/spectre-meltdown-checker/redhat_canonical/spectre-meltdown-checker.sh ).

About the Amazon kernels, I don't know if they have their own version of the mitigation, different from the upstream, Canonical and RedHat ones, or not. If the above branch doesn't yield better results, would you be able to upload one of those kernels somewhere for me to have a look ? Thanks !

**Comment by SrikanthSoma** [comment]
Thanks for your quick response. I cloned new branch redhat_canonical and ran the script on following OS ubuntu14.04,ubuntu16.04,redhat6.9,redhat7.x

Ubuntu14.04 kernel version: 3.13.0-141-generic and Ubuntu16.04 kernel version: 4.4.0-1049-aws according to ubuntu official documentation it's patched but script says variant1 as unknown, variant2 as vulnerable, variant3 as not vulnerable

For redhat6.9 kernel version: 2.6.32-696.20.1.el6.x86_64
,redhat7.x kernel version: 3.10.0-327.36.1.el7.x86_64 it gives accurate information

**Comment by speed47** [comment]
I tried a several kernels on this branch, including the `Ubuntu 3.13.0-141.190-generic`, is this the exact version you've tested or is this a slightly different one? Could you paste the entire output of the script in verbose mode, so we can find out why you're having different results?

About the `4.4.0-1049-aws`, I assume from the name it's not an official Ubuntu kernel but an Amazon one, I couldn't test on any Amazon kernel, do you know if it's available for download somewhere?

Thanks!

**Comment by SrikanthSoma** [comment]
<img width="1425" alt="ubuntu14 04" src="https://user-images.githubusercontent.com/17818558/36284547-e3ebbe3c-125c-11e8-9e30-31613b076829.png">

**Comment by speed47** [comment]
As shown in the script, you should install some packages for correct variant1 detection : `binutils` and  `strings`

**Comment by speed47** [comment]
For variant2, you are indeed vulnerable because your CPU microcode doesn't support IBRS yet (even if your kernel does). Intel is currently releasing updates, so depending on your CPU model, it should be in the following days/weeks
